### PR TITLE
Adding Stream -> HAL conversion.

### PR DESCRIPTION
**Pull request by benvanik**
This is being added in parallel with the existing Flow -> HAL conversion so that we can still guard streams behind a flag. Once the flag is flipped the existing HAL/Conversion/ code will be deleted as well as most of the HAL passes that now are handled in the Stream dialect. MaterializeInterfaces was the only pass that required reworking insofar as much of the current one could be deleted (all the constant pool handling and binding simplification). Much of the code in here is transitory pending the large HAL dialect cleanup (removing awareness of tensors/etc).

One of the larger changes here during conversion is that tensors that survive to the HAL dialect are converted directly to buffer views - prior to this they were buffers which then required a bunch of shady stuff to recalculate the tensor parameters. The TensorRewriteAdaptor will no longer be needed, and tensor->buffer_view is just a type change 🥳 

This adds the `-iree-experimental-streams` flag to iree-translate which can be used to enable the new codepath. For this PR it defaults to disabled.

### CODE REVIEW

File: iree/compiler/Dialect/HAL/Transforms/ConvertToHAL2.cpp [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    // Gather all HAL dialect conversion patterns from custom dialects.
+    // These will perform the tensor->buffer mapping for their ops.
+    for (auto *conversionInterface : conversionInterfaces) {
+      conversionInterface->setupConversionTarget(conversionTarget, patterns,
+                                                 typeConverter);
+    }
+
+    // NOTE: we allow ops that we don't know about to allow custom dialects
+    // that don't need anything HAL-specific to pass through. This is handled by
+    // the fallback type legality support of the
```
Comment (ScottTodd) [inline iree/compiler/Dialect/HAL/Transforms/ConvertToHAL2.cpp:94]: "support of the" what? :p

File: iree/compiler/Dialect/HAL/Conversion2/StreamToHAL/ConvertStreamToHAL.cpp [hunk 2]
```diff
@@ -0,0 +1,1149 @@
...
+          break;
+        case IREE::Stream::Lifetime::External:
+          // #yolo; these come from/go to outside the program.
+          // Today we assume they are device-local|host-visible just for
+          // practical purposes but that does not have to be true. We really
+          // want this to be something we analyze and handle on the edges
+          // (transfering devices/etc if needed).
+          memoryTypes = memoryTypes |
+                        IREE::HAL::MemoryTypeBitfield::DeviceLocal |
+                        IREE::HAL::MemoryTypeBitfield::HostVisible;
```
Comment (ScottTodd) [inline iree/compiler/Dialect/HAL/Conversion2/StreamToHAL/ConvertStreamToHAL.cpp]: If the memory is not `device-local|host-visible` what will happen? Anything we can do now to give a better error message until this is generalized?

File: iree/compiler/Dialect/HAL/Conversion2/StreamToHAL/ConvertStreamToHAL.cpp [hunk 3]
```diff
@@ -0,0 +1,1149 @@
...
+    : public StreamConversionPattern<IREE::Stream::TimepointJoinOp> {
+  using StreamConversionPattern::StreamConversionPattern;
+  LogicalResult matchAndRewrite(
+      IREE::Stream::TimepointJoinOp joinOp, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    // TODO(benvanik): model timepoints as semaphores.
+    // This should be a max() of the operand timepoints. MLIR has no max op
+    // until https://reviews.llvm.org/D110540. Could be done with affine
+    // expressions, but since everything is always 0 we just max(0,0)=0 here :)
+    rewriter.replaceOpWithNewOp<arith::ConstantIndexOp>(joinOp, 0);
```
Comment (ScottTodd) [inline iree/compiler/Dialect/HAL/Conversion2/StreamToHAL/ConvertStreamToHAL.cpp:1056]: The linked https://reviews.llvm.org/D110540 has landed, FYI. (fine to address in a follow-up)

**Review by ScottTodd — APPROVED** [review]

File: iree/compiler/Dialect/HAL/Conversion2/StreamToHAL/ConvertStreamToHAL.cpp (hunk 2 above)
Comment (benvanik) [reply iree/compiler/Dialect/HAL/Conversion2/StreamToHAL/ConvertStreamToHAL.cpp]: the hal.buffer.assert op/export added after this give a nice message here on input if the memory type or usage bits don't match 🎉

File: iree/compiler/Dialect/Util/Analysis/Explorer.cpp [hunk 4]
```diff
@@ -24,6 +24,26 @@ static StringRef getRegionName(Region &region) {
...
 }
 
+// Returns the remapped successor operand index if the branch operand is
+// passed to a successor (vs being used by the op itself, such as the cond_br
+// condition).
+static llvm::Optional<unsigned> mapSuccessorOperand(BranchOpInterface branchOp,
+                                                    unsigned successorIdx,
+                                                    unsigned operandIdx) {
+  // I don't know if there's a better way to do this - the interface doesn't
+  // help.
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/compiler/Dialect/Util/Analysis/Explorer.cpp:34]: Yeah, the interface could have this method on it (I feel like I've written this helper at least twice during the streams work). The interface needs some love in general to be more usable. There are definitely some gaps in here that need some more testing with respect to scf.for & co and we'll need to flesh out along with several other parts of the system once we have more of those. The current flow->hal path doesn't support them at all and the new streams path _mostly_ supports them, but it'll be easier to start plumbing through the ops and not baking them into CFG at the input side once we can eject the old path.

**Pull request merged**